flytepropeller/pkg/controller/nodes/attr_path_resolver.go

```diff
@@ -0,0 +1,155 @@
+func resolveAttrPathInPromise(nodeID string, literal *core.Literal, bindAttrPath []*core.PromiseAttribute) (*core.Literal, error) {
+	var currVal *core.Literal = literal
+	var tmpVal *core.Literal
+	var err error
+	var exist bool
+	count := 0
+
+	for _, attr := range bindAttrPath {
+		switch currVal.GetValue().(type) {
+		case *core.Literal_Map:
+			tmpVal, exist = currVal.GetMap().GetLiterals()[attr.GetStringValue()]
+			if !exist {
+				return nil, errors.Errorf(errors.PromiseAttributeResolveError, nodeID, "key [%v] does not exist in literal %v", attr.GetStringValue(), currVal.GetMap().GetLiterals())
+			}
+			currVal = tmpVal
+			count++
+		case *core.Literal_Collection:
+			if int(attr.GetIntValue()) >= len(currVal.GetCollection().GetLiterals()) {
+				return nil, errors.Errorf(errors.PromiseAttributeResolveError, nodeID, "index [%v] is out of range of %v", attr.GetIntValue(), currVal.GetCollection().GetLiterals())
+			}
+			currVal = currVal.GetCollection().GetLiterals()[attr.GetIntValue()]
+			count++
+		// scalar is always the leaf, so we can break here
+		case *core.Literal_Scalar:
+			break
+		}
+	}
+
+	// resolve dataclass
+	if currVal.GetScalar() != nil && currVal.GetScalar().GetGeneric() != nil {
+		st := currVal.GetScalar().GetGeneric()
+		// start from index "count"
+		currVal, err = resolveAttrPathInPbStruct(nodeID, st, bindAttrPath[count:])
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return currVal, nil
+}
+
+// resolveAttrPathInPbStruct resolves the protobuf struct (e.g. dataclass) with attribute path
+func resolveAttrPathInPbStruct(nodeID string, st *structpb.Struct, bindAttrPath []*core.PromiseAttribute) (*core.Literal, error) {
+
+	var currVal interface{}
+	var tmpVal interface{}
+	var exist bool
+
+	currVal = st.AsMap()
+
+	// Turn the current value to a map so it can be resolved more easily
+	for _, attr := range bindAttrPath {
+		switch resolvedVal := currVal.(type) {
+		// map
+		case map[string]interface{}:
+			tmpVal, exist = resolvedVal[attr.GetStringValue()]
+			if !exist {
+				return nil, errors.Errorf(errors.PromiseAttributeResolveError, nodeID, "key [%v] does not exist in literal %v", attr.GetStringValue(), currVal)
+			}
+			currVal = tmpVal
+		// list
+		case []interface{}:
+			if int(attr.GetIntValue()) >= len(resolvedVal) {
+				return nil, errors.Errorf(errors.PromiseAttributeResolveError, nodeID, "index [%v] is out of range of %v", attr.GetIntValue(), currVal)
+			}
+			currVal = resolvedVal[attr.GetIntValue()]
+		}
+	}
+
+	// After resolve, convert the interface to literal
+	literal, err := convertInterfaceToLiteral(nodeID, currVal)
+
+	return literal, err
+}
+
+// convertInterfaceToLiteral converts the protobuf struct (e.g. dataclass) to literal
+func convertInterfaceToLiteral(nodeID string, obj interface{}) (*core.Literal, error) {
+
+	literal := &core.Literal{}
+
+	switch obj := obj.(type) {
+	case map[string]interface{}:
+		newSt, err := structpb.NewStruct(obj)
```

Comment:
just one question.  if you just have a typing.Dict[str, DataClass], and you bind to one of the dataclasses.  you end up in this condition right?

What is this line for?  `newSt, err := structpb.NewStruct(obj)`

In fact in this case - line 22 above:
`tmpVal, exist = currVal.GetMap().GetLiterals()[attr.GetStringValue()]`
won't tmpVal already be a Literal?  can't you just return it?  it won't get returned right now, it'll fall into the condition on line 41.

Reply:
i guess you are talking about the case that where `bindAttrPath[count:]` is 0 and the `currVal` on line41 is a datatclass. Currently we convert convert struct to map, and convert it back to struct without doing other things because line 63 the loop will be skipped. 

adding a check about `len(bindAttrPath[count:]) == 0` on line 41 can circumvent the unnecessary work, but I think it's ok to keep it as-is because it should not be a large overhead.